Add federated repositories

cawilson opened the pull request:
- [x] All [tests](https://github.com/jfrog/jfrog-client-go#tests) passed. If this feature is not already covered by the tests, I added new tests.
- [x] This pull request is on the dev branch.
- [x] I used gofmt for formatting the code before submitting the pull request.
-----

Added all supported `federated repository` types, this is required before adding the necessary Terraform resources in https://github.com/jfrog/terraform-provider-artifactory, which is a much needed feature before we can switch to using `federated repositories`.

cawilson commented:
I have read the CLA Document and I hereby sign the CLA

cawilson commented:
👋🏻 @eyalbe4 any chance for feedback on this PR if you have the time?

eyalbe4 reviewed (changes requested):
Thanks so much for this great contribution!
See my inline suggestions,

cawilson commented on artifactory/services/federatedrepository.go:
```diff
@@ -0,0 +1,448 @@
+	Enabled *bool  `json:"enabled,omitempty"`
+}
+
+type FederatedRepositoryBaseParams struct {
```
@eyalbe4 yeah, looks like there is a lot of duplication in the structs across all the repository types. Are you happy for me to deduplicate across all the repository types or just focus on this one for now? I do want to fill in the gaps of the other repository types as well eventually, so it's either now or later I suppose.

eyalbe4 replied:
@cawilson - I think you can go ahead and remove the duplicated structs across all repository types now.

cawilson replied:
@eyalbe4 Sounds good. I'll make the changes an update this PR to include them.

cawilson commented on artifactory/services/federatedrepository.go:
```diff
@@ -0,0 +1,448 @@
+	return IvyFederatedRepositoryParams{FederatedRepositoryBaseParams: FederatedRepositoryBaseParams{Rclass: "federated", PackageType: "ivy"}}
+}
+
+type CommonMavenGradleFederatedRepositoryParams struct {
```
@eyalbe4 there are common properties that apply to `maven`, `gradle`, `ivy`, and `sbt` and I'm struggling to come up with a good name instead of `CommonMavenGradleIvySbtRepositoryParams` but I'm unsure what name could be used to define them all. Do you have any ideas or are you happy to just keep adding the package names to the struct name?

eyalbe4 replied:
It's definitely not a perfect name, but the only name I can come up with right now is ```JavaPackageManagersRepositoryParams```.

cawilson replied:
Done ✔️

cawilson commented:
@eyalbe4 - Sorry this took so long, life/work got in the way. I've updated the PR with additional changes based on your suggestions but took it a bit further by:
- De-duplicated all the params across all repository/package types
- De-duplicated the repeated function call to create/update repositories
- Filled in missing params across all repository/packages
- Added missing package types to all repository types
- Re-ordered tests and repositories in alphabetical order
- Updated all repository tests
  - Removed duplication
  - All local, remote, federated and virtual tests are passing
- Included new function to convert a local to a federated repository
  - Added tests, passing
  - Added example to readme

Sorry, that's a lot to review and I hope it's all ok. Please let me know what you think.

cawilson commented:
@eyalbe4 - Also added functions for triggering full federated repository synchronisation, which is important when converting local repositories for the first time.

eyalbe4 commented on README.md:
````diff
@@ -693,6 +690,49 @@ Updating virtual Go repository:
 err = servicesManager.UpdateVirtualRepository().Go(params)
 ```
 
+#### Creating and Updating Federated Repository
````
Let's add all the new README section to the table of content as well.

cawilson replied:
Oh yes, I missed that. I've updated the table of contents but just thinking, now that GitHub has a table of contents built in now, this is probably no longer needed and reduces some maintenance on the README section.

eyalbe4 commented on README.md:
```diff
@@ -693,6 +690,49 @@ Updating virtual Go repository:
+
+Alpine, Bower, Cran, Cargo, Chef, Cocoapods, Composer, Conan, Conda, Debian, Docker, Gems, Generic, Gitlfs, Go, Gradle, Helm, Ivy, Maven, Npm, Nuget, Opkg, Puppet, Pypi, Rpm, Sbt, Vagrant and Yum
+
+Each package type has it's own parameters struct, can be created using the method
```
it's --> its
(looks like this typo was copied from somewhere elase in the README, if you're able to fix all, that would be great :))

cawilson replied:
Done ✔️ (including the other 3 locations)

cawilson commented on tests/utils.go:
```diff
@@ -626,10 +763,13 @@ func createRepoConfigValidationFunc(repoKey string, expectedConfig interface{})
+			// The password field may be encrypted and won't match the value set, need to handle this during validation
+			if key == "password" && len(fmt.Sprintf("%s", expectedValue)) > 0 && len(fmt.Sprintf("%s", confMap[key])) > 0 {
+				continue
+			}
```
Done ✔️

cawilson commented on artifactory/manager.go:
```diff
@@ -396,6 +414,24 @@ func (sm *ArtifactoryServicesManagerImp) GetReplication(repoKey string) ([]utils
 	return getPushReplicationService.GetReplication(repoKey)
 }
 
+func (sm *ArtifactoryServicesManagerImp) ConvertLocalToFederated(repoKey string) error {
```
Done ✔️

eyalbe4 reviewed (changes requested):
```
# github.com/jfrog/jfrog-client-go/artifactory
artifactory/emptymanager.go:387:5: cannot use (*EmptyArtifactoryServicesManager)(nil) (type *EmptyArtifactoryServicesManager) as type ArtifactoryServicesManager in assignment:
	*EmptyArtifactoryServicesManager does not implement ArtifactoryServicesManager (missing ConvertLocalToFederated method)
artifactory/manager.go:55:2: cannot use manager (type *ArtifactoryServicesManagerImp) as type ArtifactoryServicesManager in return argument:
	*ArtifactoryServicesManagerImp does not implement ArtifactoryServicesManager (missing ConvertLocalToFederated method)
```

cawilson commented:
> ```
> # github.com/jfrog/jfrog-client-go/artifactory
> artifactory/emptymanager.go:387:5: cannot use (*EmptyArtifactoryServicesManager)(nil) (type *EmptyArtifactoryServicesManager) as type ArtifactoryServicesManager in assignment:
> 	*EmptyArtifactoryServicesManager does not implement ArtifactoryServicesManager (missing ConvertLocalToFederated method)
> artifactory/manager.go:55:2: cannot use manager (type *ArtifactoryServicesManagerImp) as type ArtifactoryServicesManager in return argument:
> 	*ArtifactoryServicesManagerImp does not implement ArtifactoryServicesManager (missing ConvertLocalToFederated method)
> ```

Whoops, missed the interface declaration when renaming. Should be better now 😄

eyalbe4 commented:
Thanks so muc for this awesome contribution @cawilson!
v1.3.0 which includes this enhancement is now released.

cawilson commented:
> Thanks so muc for this awesome contribution @cawilson!
> v1.3.0 which includes this enhancement is now released.

Thankyou @eyalbe4 for assisting with this PR and contribution 😄
